gadopt/momentum_equation.py

```diff
@@ -47,195 +42,147 @@ class ViscosityTerm(BaseTerm):
+                * sigma
+                * inner(outer(eq.n, eq.test), mu * trial_tensor_jump)
+                * eq.ds(bc_id)
+            )
+            F -= inner(mu * nabla_grad(eq.test), trial_tensor_jump) * eq.ds(bc_id)
+            F -= inner(outer(eq.n, eq.test), stress) * eq.ds(bc_id)
+
+        if "un" in bc:
+            trial_tensor_jump = outer(eq.n, eq.n) * (dot(eq.n, trial) - bc["un"])
+            if compressible_stress:
+                trial_tensor_jump -= 2 / 3 * identity * (dot(eq.n, trial) - bc["un"])
+            trial_tensor_jump += transpose(trial_tensor_jump)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma
+                * inner(outer(eq.n, eq.test), mu * trial_tensor_jump)
+                * eq.ds(bc_id)
+            )
+            F -= inner(mu * nabla_grad(eq.test), trial_tensor_jump) * eq.ds(bc_id)
+            # We only keep the normal part of stress; the tangential part is assumed to
+            # be zero stress (i.e. free slip) or prescribed via "stress".
+            F -= dot(eq.n, eq.test) * dot(eq.n, dot(stress, eq.n)) * eq.ds(bc_id)
+
+        if "stress" in bc:  # a momentum flux, a.k.a. "force"
+            # Here we need only the third term because we assume jump_u = 0
+            # (u_ext = trial) and stress = n . (mu . stress_tensor).
+            F -= dot(eq.test, bc["stress"]) * eq.ds(bc_id)
+
+        if "normal_stress" in bc:
+            F += dot(eq.test, bc["normal_stress"] * eq.n) * eq.ds(bc_id)
+
+    return -F
+
+
+def pressure_gradient_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.test)
+
+    F = -dot(div(eq.test), eq.p) * eq.dx
+
+    # Integration by parts gives a natural condition on pressure (as part of a normal
+    # stress condition). For boundaries where the normal component of u is specified, we
+    # remove that condition.
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc or "un" in bc:
+            F += dot(eq.test, eq.n) * eq.p * eq.ds(bc_id)
+
+    return -F
+
+
+def divergence_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.u)
+
+    rho = eq.rho_mass
+    F = -dot(eq.test, div(rho * eq.u)) * eq.dx
+
+    # Add boundary integral for bcs that specify the normal component of u.
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc:
+            F -= eq.test * rho * dot(eq.n, bc["u"] - eq.u) * eq.ds(bc_id)
+        elif "un" in bc:
+            F -= eq.test * rho * (bc["un"] - dot(eq.n, eq.u)) * eq.ds(bc_id)
+
+    return -F
+
+
+def momentum_source_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    source = eq.approximation.buoyancy(eq.p, eq.T) * upward_normal(eq.mesh)
+    F = -dot(eq.test, source) * eq.dx
+
+    return -F
+
+
+viscosity_term.required_attrs = {"u", "stress"}
```

Comment:
as mentioned in stokes integrators, is 'u' a required attribute for viscosity given its the trial function?

Reply:
Good catch, thank you.